feat : [001-signUp,signIn]

karpei-taemukan opened the pull request:
<div align="center"> <h2>개발한 기능의 진행 순서</h2> </div>
<br>
<div align="center">
<div>
 회원 가입(mailgun 통해서 인증 메일 발송)
<div align="center">
※ mailgun을 사용하기 때문에 메일 테스트를 하실려면 mailgun에 가입하여, 도메인을 만들고, api 키를 얻어야 합니다 </div>
</div>
<br>
<div>
 ⬇
</div>
<br>
<div>
인증 메일을 받은 사용자가 이메일 함에서 인증 메일의 인증 url 클릭(클릭하면 인증 완료(GetMapping 으로 구현))
</div>
<br>
<div>
 ⬇
</div>
<br>
<div>
로그인 하기 (이 단계에서 JWT토큰이 발행됩니다)
</div>
<br>
<div>
 ⬇
</div>
<br>
<div>
로그인 후,  계좌 가입 (계좌는 최대 5개까지만 가입 가능 )
</div>
</div>
<br><br>
<div align="center"> <h2>PR을 해주셨으면 하는 부분</h2> </div>
<br>
<div align="center">
<div>
1.  불필요한 코드나 개선할 수 있는 코드가 있는 지에 대해 의견 남겨주셨으면 합니다!
</div>
<br><br>
<div>
2.  혹여나 잘못된 논리의 코드가 있는 지에 대한 의견 남겨주셨으면 합니다!
</div>
<br><br>
<div>
3.  조금 더 나은 방식이 있다면 꼭 의견 남겨주셨으면 합니다!
</div>
<br><br>
<div>
4. 그 외에 피드백 주실 게 있다면 부담 없이 의견 남겨주셨으면 합니다!
</div>
<br><br>
</div>

karpei-taemukan committed:
feat : [001-signUp,signIn]
 1. 사용자에 대한 회원 가입
 2. 회원 가입 후 mailgun을 통해 인증메일 보내기
 3. 로그인 기능 --> 로그인 후 JWT 토큰 발행
 4. 계좌 가입(로그인을 해서 JWT 토큰 발행 후 계좌 가입 가능)

spring-kotlin-lover commented on src/main/java/com/zerobase/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,13 @@
+public class GlobalExceptionHandler {
+  @ExceptionHandler(AccountException.class)
+  public ErrorResponse handleAccountUserException(AccountException e){
+    return new ErrorResponse(e.getErrorCode(), e.getErrorMessage());
```
error 상황별로 http status 를 정의해보는 것은 어떨까요? 지금 현재 구현된 사항으로는 항상 200 을 응답해주는 것으로 되어 있네요.

error 상황별로 어떤 http status 가 적합할지에 대해 한번 생각해보시고, 각 상황에 맞는 status 를 내려주도록 정리해보시면 좋을 듯 합니다.

spring-kotlin-lover replied:
error code 에 http status 를 정의하도록 개선해주셨는데요. 아쉽게도 이는 과거 코드와 달라진 것이 없습니다.

ErrorResponse 객체를 리턴하게 되면, 200 응답을 항상 내려주기 떄문이에요. ResponseEntity 객체 기반으로 http status 를 정의해서 내려주도록 개선해야될듯 싶습니다.

ErrorResponse 에는 Http status 가 내려가지 않아도 될 것으로 보입니다. 올바른 http status 가 내려가면 이미 client 에서도 http status 기반으로 인지할 수 있기 때문입니다.


한번 테스트 코드를 작성해보는 것까지 시도해보면 어떨까요? ^^; 내가 작성한 코드에 대한 기대값을 검증하지 못하여 제대로 개선이 안된듯 합니다.

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -0,0 +1,193 @@
+
+
+
+  public boolean isEmailExist(String email){
+    return accountUserRepository.findByEmail(email.toLowerCase(Locale.ROOT)).isPresent();
+  }
+
+  private boolean isAccountExist(String accountName) {
+    return accountRepository.findByAccountName(accountName.toLowerCase(Locale.ROOT)).isPresent();
+  }
```
exist 여부에 대한 메서드라면, jpa 에서 제공하는 `exists` 쿼리를 적용해보는 것은 어떨까요? 현재 작성해주신 것과 큰 성능차이는 없을 것으로 예상되지만 한번 공부해보시면 좋을 듯 합니다~

https://velog.io/@hyunho058/JPA-Exists

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -0,0 +1,193 @@
+  }
+
+
+//----------------------------------------------------------------------------------------
```
불필요한 주석은 항상 삭제해주도록 노력해주시기를 부탁드립니다.

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -0,0 +1,193 @@
+          .text(getVerificationEmailBody(form.getEmail(), form.getName(), code))
+          .build();
+
+      System.out.println("EMAIL FORM :"+sendMailForm);
```
console 로 출력하는 것보다는, logger 를 활용해보는 것은 어떨까요?  실제 서비스하면서 참고할만한 데이터들은 로그를 찍어 디버깅을 용이하게 하고 있습니다.

더 나아가서는 로그레벨에 따라 어떤 로그를 남길 것인지도 고민하게 됩니다.
https://dkswnkk.tistory.com/445

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -0,0 +1,193 @@
+    AccountUser accountUser =  accountUserRepository.findByEmail(email)
+        .orElseThrow(() -> new AccountException(ErrorCode.USER_NOT_FOUND));
+
+    if(!accountUser.getVerificationCode().equals(code)){
```
```suggestion
    if (!accountUser.getVerificationCode().equals(code)) {
```
coding convention 에 대해서는 관련 내용을 확인해보시고, 항상 숙지해보는 습관을 들이면 좋을 것 같아요.
실제 현업에서는 개발자들간의 coding style 을 통일하여 전체적인 style 을 맞추어 혼란을 줄이려는 노력을 하고 있습니다.
https://naver.github.io/hackday-conventions-java/

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -0,0 +1,193 @@
+      validateCreateAccount(accountUser);
+
+
+      String accountNumber = generateAccountNumber();
```
작성해주신 로직이 랜덤 기반으로 구현되어 있는데요, 낮은 확률로 계좌 번호가 중복되면 어떻게 될까요?

karpei-taemukan committed:
refactor : [001-signUp,signIn]
1. error 상황별로 http status 를 정의했습니다
2.  jpa 에서 제공하는 exists 쿼리를 적용했습니다
3. 불필요한 주석은 제거하고 coding convention 을 적용했습니다
4. log 를 활용해 로그를 남겼습니다
5. 랜덤 기반 계좌 번호 생성에서 중복된 계좌 번호 생성시, 예외를 발생시켜, 사용자로 하여금 다시 계좌 가입을 요청하도록 메세지를 전달했습니다

spring-kotlin-lover commented on src/main/java/com/zerobase/service/SignupService.java:
```diff
@@ -145,46 +130,51 @@ public AccountDto accountSignUp(String token, SignUpAccountForm form) {
+    Integer accountCount = accountRepository.countByAccountNumber(accountNumber);
 
-      String accountNumber = generateAccountNumber();
+    // 생성한 계좌 번호가 이미 있는 경우
+    if (accountCount > 0) {
+      throw new RandomCodeException(ErrorCode.DUPLICATED_RANDOM_CODE);
+    }
```
계좌번호를 생성하는 것은 client 의 책임이 아닌 server 의 책임으로 보입니다. 이 경우에 exception 을 내려주는 것이 아니라, 중복되지 않는 계좌번호를 생성할 때까지 지속적으로 해주어야 될 것으로 보이네요.
계좌번호를 client 가 보낸 것이 아니라, client 입장에서는 황당하기 떄문이죠.

서버측에서 중복되지 않은 계좌번호를 어떻게 하면 잘 만들어볼 수 있을까요? 설계적으로 조금 더 고민해보면 좋을 듯 합니다.

karpei-taemukan committed:
refactor : [002-random-code]

1. 중복된 계좌 번호가 있을 경우에 예외를 던지는 로직을 삭제했습니다

2. 생성한 계좌번호가 이미 있는 경우 재귀호출로 계좌 번호 다시 생성하도록 하였습니다

karpei-taemukan committed:
Update SignupService.java

karpei-taemukan committed:
refactor : [003-change Exception Handler]
@ResponseStatus 으로 exception 응답을 주는 방식에서
ResponseEntity 객체로 exception 응답을 주는 방식 바꿨습니다

karpei-taemukan committed:
Merge remote-tracking branch 'origin/feature/signup' into feature/signup

# Conflicts:
#	src/main/java/com/zerobase/service/SignupService.java
